R/compare_diagnoses.R

```diff
@@ -0,0 +1,291 @@
+  
+  
+  # Function to split bootstrap into groups defined by merge by set
+  split_bootrstrap <- function(diagnosis, group_by_set = merge_by_set){
+    bootstrap_df <- diagnosis$bootstrap_replicates
+    group_by_list <- bootstrap_df[, group_by_set, drop = FALSE]
+    split(bootstrap_df, lapply(group_by_list, addNA), drop = TRUE)
+  }
+  bootstrap_df1 <- split_bootrstrap(diagnosis1)
+  bootstrap_df2 <- split_bootrstrap(diagnosis2)
+  
+  # Function to create  output.
+  # For each combination of estimand_label, estimator_label and term in comparison_list
+  # take the mean and the se of each estimated diagnosand "d" from comparison_list
+  # and unite with 
+  # the mean difference computed as mean(bootstrap_df2[,"d"] - bootstrap_df1[, "d"]
+  # se of the difference computed as sd(bootstrap_df2[,"d"] - bootstrap_df1[, "d"])
+  # p.value computed as the proportion of times that the difference is on opposite side of zero from mean difference 
+  create_difference_df <- function(label, 
+                                   df1 = bootstrap_df1, 
+                                   df2= bootstrap_df2,
+                                   comparison = comparison_list,
+                                   diagnosands_list = diagnosands) {
+    # Estimated diagnosands and se(diagnosands) from each design
+    d1 <- df1[[label]]
+    d2 <- df2[[label]]
+    comp <- comparison[[label]]
+    c_names <- colnames(comp)
+    set <-   grepl("^estimand_label", c_names)| grepl("^estimator_label", c_names)| grepl("^term", c_names) 
+    select_cols <- c_names[grepl("^design_label", c_names)]
+    select_cols <- c(select_cols, c_names[set])
+    identifiers <- comp[,   select_cols]
+    
+    # Compute difference stats
+    difference <- do.call(rbind, lapply(diagnosands_list, function(diagnosand){
+      diagnosand1 <- comp[,paste0(diagnosand, "_1")]
+      diagnosand2 <- comp[,paste0(diagnosand, "_2")]
+      se1 <- comp[,paste0("se(",diagnosand, ")_1")]
+      se2 <- comp[,paste0("se(",diagnosand, ")_2")]
+      diff <- d2[, diagnosand] - d1[, diagnosand] 
+      mu <- mean(diff, na.rm = TRUE) 
+      conf.low  <- quantile(diff, (1 - level)/2, na.rm = TRUE)
+      conf.high <- quantile(diff, level + (1 - level)/2 , na.rm = TRUE)
+
+      # Combine
+     out<- data.frame(diagnosand =diagnosand,
+                 mean_1 = diagnosand1,
+                 mean_2 = diagnosand2,
+                 mean_difference = mu ,
+                 se_1 = se1,
+                 se_2 = se2,
+                 se_difference = sd(diff),
+                 conf.low  = conf.low,
+                 conf.high = conf.high,
+                 sims = sims, 
+                 bootstrap_sims = bootstrap_sims,
+                stringsAsFactors = FALSE)
+     
+      }))
+    suppressWarnings(cbind(identifiers, difference))
+
+  }
+  
+  return_df <- lapply(names(comparison_list),  create_difference_df )
+  return_df <- data.frame(rbind_disjoint(return_df, infill = ""))
```

Comment:
just a question -- is this call to data.frame required?  I thought rbind_disjoint always returned a data.frame, but I could be wrong

Reply:
AFAICT rbind_disjoint always returns a data.frame.